library/lua/gui.lua

```diff
@@ -955,19 +983,35 @@ INTERIOR_MEDIUM_FRAME = FRAME_INTERIOR_MEDIUM
 function paint_frame(dc,rect,style,title,inactive,pause_forced,resizable)
     local pen = style.frame_pen
     local x1,y1,x2,y2 = dc.x1+rect.x1, dc.y1+rect.y1, dc.x1+rect.x2, dc.y1+rect.y2
-    dscreen.paintTile(style.lt_frame_pen or pen, x1, y1)
-    dscreen.paintTile(style.rt_frame_pen or pen, x2, y1)
-    dscreen.paintTile(style.lb_frame_pen or pen, x1, y2)
-    local rb_frame_pen = style.rb_frame_pen
-    if rb_frame_pen == FRAME_WINDOW.rb_frame_pen and not resizable then
-        rb_frame_pen = FRAME_PANEL.rb_frame_pen
+    local signed = false
+
+    if x1 == x2 and y1 == y2 then
+        dscreen.paintTile(style.x_frame_pen or pen, x1, y2)
+    elseif x1 == x2 then
+        dscreen.paintTile(style.tte_frame_pen or style.tti_frame_pen or pen, x1, y1)
```

Comment:
it does not seem convenient to get an "interior"-style border with this logic. a frame would have to remove all the `t*e` fields so this logic would fall through to using the `t*i` fields.

what if:
- the border just had one set of t-junction variables (and maybe capitalize the `T` to distinguish it from "top"? e.g. `tT_frame_pen`, `bT_frame_pen`, etc.)
- we choose which t-junction tiles to use in `make_frame` based on an `interior` parameter
 would that work?

Reply:
oh you mean the `style.tte_frame_pen or style.tti_frame_pen`?

yeah, like, i put those `or style.tti_frame_pen` there mostly to have this conversation. 🤣 
written like that, it would require what you described. Which is of course inconvenient af.

I too thought about adding some sort of "interior" flag or something, but i wanted to ask you if we could even do that. 🤔 what would adding such a flag entail, for the rest of dfhack? 
(like-- what would we break, what would we have to update?)

> what if:
> 
>     * the border just had one set of t-junction variables (and maybe capitalize the `T` to distinguish it from "top"? e.g. `tT_frame_pen`, `bT_frame_pen`, etc.)
> 
>     * we choose which t-junction tiles to use in `make_frame` based on an `interior` parameter
>       would that work?

This is the point i wanted to get to: 
i think `make_frame` should output a FRAME with both internal and external Ts, and *then* the "interior" flag should be settable from each Panel individually. 🤔 

Meaning we won't end up with FRAME_MEDIUM_INTERNAL and FRAME_MEDIUM_EXTERNAL, but just a FRAME_MEDIUM that could be set to use both "internal" or "external" sprites based on an additional field in the Panel/Widget or whatever. 🤔 on a case-by-case basis.

Reply:
additionally, great idea about using capital Ts to differentiate from t(op). Will definitely do that 👍

Reply:
> Meaning we won't end up with FRAME_MEDIUM_INTERNAL and FRAME_MEDIUM_EXTERNAL, but just a FRAME_MEDIUM that could be set to use both "internal" or "external" sprites based on an additional field in the Panel/Widget or whatever.  on a case-by-case basis.

what are the pros and cons of each option, as you see it?

Reply:
let's see... 

### Approach 1: a single FRAME_X per border texture
#### Pros
- a FRAME would contain a reference to \*all\* sprites in the spritesheet.
(imo this is very important, because it makes a FRAME a "complete set". Meaning that if for whatever reason you need sprite X, you know it's in there and that you can access it.)

#### Cons
- a FRAME would have 20 sprite references instead of 16, making it a bit "bigger" in memory than approach 2, i guess?
- we will have to implement some logic for choosing which sprites to use, which does not exist yet. 🤔

-------

### Approach 2: two FRAME_X_INTERNAL / FRAME_X_EXTERNAL per border texture
#### Pros
- probably easier to implement? 
- a FRAME would have up to 16 sprite references in it, making the whole structure smaller i guess?

#### Cons
- no FRAME would contain all sprites of a border spritesheet. 
- you need to have two `make_frame()` functions, or otherwise make it do "different things"
- you have two FRAMEs where you could've had just one. 🤔 also 2\*16  > 1\*20 sprites, negating one of the pros 🤦 
- if for any reason part of your frame needs to switch from internal to external, you must come up with weird solutions.

-------
#### more considerations
in gui/launcher there's an example of a divider with internal and external ends
![Untitled-1](https://github.com/DFHack/dfhack/assets/5853667/36761109-ce61-4d95-863f-9e28450b802f)
(well, at the moment the "external" end is not even there, but you get the idea)

like, idk why, but i feel that having some way of saying:
"make a 1-tall frame from here to there, and have its rightmost end be `internal`."
would sound better than
"make a 1-tall external frame from here to halfway-there, then 
 make another 1-tall internal frame from halfway-there to there"
or... something like that. °-°


though, i must stress this-- 
these are kinda uneducated opinions of a person that does not know much at all about programming, its best practices, how to be efficient at it, and stuff like that! 😫 i'm just like... "having all the parts in one container is neat!!"

Reply:
table sizes between 16 and 20 are negligible. these frames will have 1 copy that all windows share. size won't count for pro or con here.

switching between internal and external (if we were to want to do so, and I don't really understand the case where we'd want to do this) would be easy in either option, I'd think. for option 1 we'd flip a flag and for option 2 we'd assign a different `frame-style`.

from a programming perspective, option 2 is easier. consolidating the logic in `make_frame` is easier than making dynamic rendering decisions at draw time.

it's the flexibility angle that I want to understand more. in order to handle the `gui/launcher case`, we'd need to use T junctions from multiple sprite sheets since they're different widths. We'd need to construct a frankenframe for this specific use case. possible, but not elegant. if this will be a common use case, we'd want a better solution for it.

From a design perspective, in what cases would we want a single frame to have some edges be internal and others be external?